### PR TITLE
Feature/survey

### PR DESCRIPTION
**Pull request by julsemaan**
# Description
* Adds the ability to create surveys on the captive portal
* Disables useless intermediate step in OAuth login if there is no AUP
* Capture gender when performing Google OAuth

# Impacts
Captive portal

# Delete branch after merge
YES

# NEWS file entries
## New Features
* Added the ability to create surveys in the captive portal to prompt for any informations

## Enhancements
* Disabled useless intermediate step in OAuth login if there is no AUP
* Capture gender when performing Google OAuth

### CODE REVIEW

File: html/pfappserver/lib/pfappserver/Form/Config/PortalModule/Survey.pm [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+sub BUILD {
+    my ($self) = @_;
+    $self->field('template')->default($self->for_module->meta->find_attribute_by_name('template')->default->());
+}
+
+sub options_surveys {
+    my $self = shift;
+    my @surveys = map { $_ => $_ } @{pf::ConfigStore::Survey->new->surveyIds};
```
Comment (jrouzierinverse) [inline html/pfappserver/lib/pfappserver/Form/Config/PortalModule/Survey.pm]: rewrite 
```perl
my @surveys = map { $_ => $_ } @{pf::ConfigStore::Survey->new->surveyIds};
unshift @surveys, ("" => "");
```
as
```perl
my @surveys = map { $_ => $_ } "", @{pf::ConfigStore::Survey->new->surveyIds};
```

File: docs/PacketFence_Administration_Guide.asciidoc [hunk 2]
```diff
@@ -1281,6 +1281,129 @@ The configuration is simple, you have a role which is the 'Admin role' the one a
...
+
+PacketFence has the ability to perform surveys via the captive portal and store the results in dedicated tables in the database.
+
+Setup
+^^^^^
+
+In order for the survey tables to be created automatically based on the definition of your surveys, you must grant create and alter rights to the database user defined in pf.conf. By default this user is 'pf'. On your database, connect to the MySQL CLI as root and execute the following:
+
+  mysql> GRANT CREATE,ALTER ON pf.* TO 'pf'@'%';
+  mysql> GRANT CREATE,ALTER ON pf.* TO 'pf'@'localhost';
```
Comment (jrouzierinverse) [inline docs/PacketFence_Administration_Guide.asciidoc:1295]: I would prefer not to add those privileges to the main pf user as a security issues.
Here are a list of solutions of possible alternative approaches.

* Create a dedicated user that is allowed to create and alter tables.
   That user will only be used when configuring the database.

* Have a dedicated database for the surveys that the pf user has permissions to CREATE,ALTER.

File: lib/pf/Survey.pm [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+=head2 _db_data
+
+Execute a query through pf::db
+
+=cut
+
+sub _db_data {
+    my ($class, $from_module, $module_statements_ref, $query, @params) = @_;
+
+    my $sth = db_query_execute($from_module, $module_statements_ref, $query, @params) || return ($FALSE);
```
Comment (jrouzierinverse) [inline lib/pf/Survey.pm]: Use the dal
```perl
my ($status, $sth) = pf::dal->db_execute($sql, @bind);
```

File: lib/pf/Survey.pm [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+                $result = $result->{$query_part};
+            }
+            
+            get_logger->debug("Setting survey data '$result' to field '$field'");
+
+            $response->{$field} = $result;
+        }
+    }
+
+    use Data::Dumper ; print Dumper($response);
```
Comment (jrouzierinverse) [inline lib/pf/Survey.pm]: remove

File: lib/pf/Survey.pm [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+    
+    my $sqla = SQL::Abstract::More->new();
```
Comment (jrouzierinverse) [inline lib/pf/Survey.pm]: use sql abstract through the dal
```perl
my $sqla = pf::dal->get_sql_abstract();
```

**Review by jrouzierinverse — CHANGES_REQUESTED** [review]
I do not see the form to modify the survey.conf

**Comment by julsemaan** [comment]
@jrouzierinverse 

FYI, there is no form right now since the GUI wasn't part of the scope of that work.

I guess if this picks up, we can add one but for now its all file based to interact with the survey config

File: docs/PacketFence_Administration_Guide.asciidoc (hunk 2 above)
Comment (julsemaan) [reply docs/PacketFence_Administration_Guide.asciidoc:1295]: Dedicated user might be better.

Would suffixing '_super' to the pf.conf DB user sound fair to you ?

That way its kept simple

Comment (jrouzierinverse) [reply docs/PacketFence_Administration_Guide.asciidoc:1295]: That sounds good.

Comment (jrouzierinverse) [reply docs/PacketFence_Administration_Guide.asciidoc:1295]: What will happen if I create a survey table called node?

Comment (julsemaan) [reply docs/PacketFence_Administration_Guide.asciidoc:1295]: All survey table names are always prefixed to prevent that issue

Comment (jrouzierinverse) [reply docs/PacketFence_Administration_Guide.asciidoc:1295]: OK cool

Comment (julsemaan) [reply docs/PacketFence_Administration_Guide.asciidoc:1295]: Note left in b429fab for the rights as discussed

**Comment by julsemaan** [comment]
Ready for another review

**Comment by jrouzierinverse** [comment]
rebase

**Comment by julsemaan** [comment]
It has been rebased

File: docs/PacketFence_Administration_Guide.asciidoc [hunk 6]
```diff
@@ -1281,6 +1281,137 @@ The configuration is simple, you have a role which is the 'Admin role' the one a
...
+  columns=firstname as "Firstname", lastname as "Lastname", prefered_sandwich as "Prefered Sandwich", gender as "Gender"
+
+NOTE: Make sure you reload the configuration after setting it up by doing `/usr/local/pf/bin/pfcmd configreload hard`
+
+Refer to the 'Custom Reports' section of this document for advanced configuration.
+
+Cleaning up
+^^^^^^^^^^^
+
+When you are happy with the structure of your surveys, it is recommended to remove the `CREATE` and `ALTER` rights to the `pf` database user from a security perspective. In order to do so, execute the following commands. This step is optionnal and should only be done once the structure of your survey is set in stone.
```
Comment (jrouzierinverse) [inline docs/PacketFence_Administration_Guide.asciidoc]: optionnal -> optional

**Review by jrouzierinverse — APPROVED** [review]

**Comment by julsemaan** [comment]
@jrouzierinverse 

Done

**Pull request merged**